### Vectors

Characteristics:
 - Dimensionality - number of numbers in a vector
 - Oreintation - is the vector row or col oriented
    * Generally assumed to be in col form unless specified
    * Row form vectors are written as w^T where T = transpose

Dimensions are MxN or (M, N) where M = row and N = col

In [7]:
import numpy as np


asList = [1,2,3,4]

asArray = np.array(asList)
print(asArray, '\n', asArray.shape, '\n\n')

rowVec = np.array([asList])
print(rowVec, '\n', rowVec.shape, '\n\n')

colVec = np.array([ [i] for i in asList])
print(colVec, '\n', colVec.shape, '\n\n')

[1 2 3 4] 
 (4,) 


[[1 2 3 4]] 
 (1, 4) 


[[1]
 [2]
 [3]
 [4]] 
 (4, 1) 




In [47]:
# Vector Operations

x = np.arange(0, 10)
x_reshaped = x.reshape((10, 1))

y = np.arange(10, 20)
y_reshaped = y.reshape((10, 1))

# print(x, y)

# Addition
# print(x_reshaped + y_reshaped)

# Subtraction
# print(x_reshaped - y_reshaped)

arrA = np.array([1,2,3])
arrB = np.array([4,5,6])
arrBCol = arrB.reshape((3,1))

# Broadcasting -> More Detail Later -> highly recommend this read https://www.alcala21.org/post/broadcasting/ and https://realpython.com/numpy-array-programming/
print(arrA + arrBCol, '\n')

arrC = np.array([[1,2,3]])
arrD = np.array([[10,20]]).T
print(f"{arrC}\n+\n{arrD}\n=__________________\n", arrC + arrD,'\n')


# Vector - Scalar Operations
sample_scalar = 10
print(sample_scalar * arrA, '\n')


# Vector Averaging
print(arrA + arrB * 0.5, '\n')

# Transposing
print(arrA.T.shape, arrA.shape)


# Norm -- See Notes on Norms below
print(f"arrC Norm = {np.linalg.norm(arrC)}")

[[5 6 7]
 [6 7 8]
 [7 8 9]] 

[[1 2 3]]
+
[[10]
 [20]]
=__________________
 [[11 12 13]
 [21 22 23]] 

[10 20 30] 

[3.  4.5 6. ] 

(3,) (3,)
arrC Norm = 3.7416573867739413


In [48]:
# Extra Code

# Numpy indexing https://numpy.org/doc/stable/user/basics.indexing.html

# On transposing
row_vector_1d = np.array([1, 2, 3])
print(f"Original 1D array: {row_vector_1d}, Shape: {row_vector_1d.shape}")

# Convert to a 2D column vector
column_vector_2d = row_vector_1d[:, np.newaxis]
print(f"Column vector (using newaxis): {column_vector_2d}, Shape: {column_vector_2d.shape}")

# You can then transpose this 2D column vector to a 2D row vector
row_vector_2d_from_column = column_vector_2d.T
print(f"Row vector (transposed from column): {row_vector_2d_from_column}, Shape: {row_vector_2d_from_column.shape}")

row_vector_2d = np.array([[1, 2, 3]])
print(f"Original 2D row vector: {row_vector_2d}, Shape: {row_vector_2d.shape}")

# Using the .T attribute
column_vector_from_T = row_vector_2d.T
print(f"Transposed using .T: {column_vector_from_T}, Shape: {column_vector_from_T.shape}")

# Using np.transpose() function
column_vector_from_func = np.transpose(row_vector_2d)
print(f"Transposed using np.transpose(): {column_vector_from_func}, Shape: {column_vector_from_func.shape}")

Original 1D array: [1 2 3], Shape: (3,)
Column vector (using newaxis): [[1]
 [2]
 [3]], Shape: (3, 1)
Row vector (transposed from column): [[1 2 3]], Shape: (1, 3)
Original 2D row vector: [[1 2 3]], Shape: (1, 3)
Transposed using .T: [[1]
 [2]
 [3]], Shape: (3, 1)
Transposed using np.transpose(): [[1]
 [2]
 [3]], Shape: (3, 1)


## Norms

Consider as the vector $v = [1, 2, 3]$.

### 1. L1 Norm (Manhattan Norm / Taxicab Norm)

The L1 norm is the sum of the absolute values of the vector's components.

* **Formula:** $\|v\|_1 = \sum_{i=1}^{n} |v_i|$

* **Calculation for $v = [1, 2, 3]$:**
    $\|v\|_1 = |1| + |2| + |3|$
    $\|v\|_1 = 1 + 2 + 3$
    $\|v\|_1 = 6$

### 2. L2 Norm (Euclidean Norm)

The L2 norm is the square root of the sum of the squared components. This is the most common norm and represents the standard straight-line distance.

* **Formula:** $\|v\|_2 = \sqrt{\sum_{i=1}^{n} v_i^2}$

* **Calculation for $v = [1, 2, 3]$:**
    $\|v\|_2 = \sqrt{1^2 + 2^2 + 3^2}$
    $\|v\|_2 = \sqrt{1 + 4 + 9}$
    $\|v\|_2 = \sqrt{14}$
    $\|v\|_2 \approx 3.742$

### 3. L-infinity Norm (Maximum Norm / Chebyshev Norm)

The L-infinity norm is the maximum absolute value among the vector's components.

* **Formula:** $\|v\|_\infty = \max_{i} |v_i|$

* **Calculation for $v = [1, 2, 3]$:**
    $\|v\|_\infty = \max(|1|, |2|, |3|)$
    $\|v\|_\infty = \max(1, 2, 3)$
    $\|v\|_\infty = 3$

So, for the vector $[1, 2, 3]$:
* L1 Norm = 6
* L2 Norm = $\sqrt{14} \approx 3.742$
* L-infinity Norm = 3

In [50]:
# Dot Product
a = np.array([1,2,3,4])
b = np.array([5,6,7,8])
np.dot(a,b)

np.int64(70)